### ISSUE TITLE
[Request] Better error handling when disk is full during file uploads

### ISSUE THREAD

**Issue by imitation** [OctoPrint/OctoPrint#655]
a) The "upload failed" message in a "disk full" event is the same as when uploading an invalid file: 
Upload failed
Could not upload the file. Make sure that it is a GCODE file and has the extension ".gcode" or ".gco" or that it is an STL file with the extension ".stl" and slicing support is enabled and configured.
More: 
undefined

b) Optional: There is no preemptive warning on low disk space. 

Version: 1.2.0-dev-300-g048ad78 (devel branch)

I love cookies, but i'm not sure what kind of bug report / feature request this would ideally be. So i guess it's [Misc] for now.

**Comment by dmj2x** [comment]
hello, my [partner](https://github.com/theosmacmillan) and I have had a chance to spend sometime analyzing this bug and we would like to take it if its still up for grabs.

**Comment by foosel** [comment]
Great, marking it as grabbed.

What might make things tricky here is not the writing *within OctoPrint*  to fail (so the call to the file manager, because the upload folder is "full"), but actually the file upload itself when it's still handled within tornado, due to the temporary file being written there already running into trouble.

So what I'd suggest is to first look into the possibility of determining the file size client side and match that against the free space information and directly deny the upload if that doesn't work out. But to be on the safe side (free space reported might be for a different partition than the temporary directory the upload is first done to), also have the `octoprint.server.util.tornado.UploadStorageFallbackHandler` report a specific error if there is not enough space for a file upload to go through and have the client detect that specific error and report back to the user.

To summarize:

  * There are two places where stuff can go wrong, uploading to the temporary directory still within the tornado handler of the upload request and moving the uploaded file to the upload folder within OctoPrint.
  * Client should check whether there's even a chance to upload the file fully before attempting the request.
  * Server should report some meaningful error code back if the upload fails either in the tornado part or the OctoPrint API part due to space running out (while it's actually a WebDav code instead of pure HTTP, I'd lean towards using `507 Insufficient Storage` for this), client should display meaningful error in such a case (API docs for the upload endpoint need to be extended to specify the new error code too)

**Comment by theosmacmillan** [comment]
I'm working with David [dmj2x](https://github.com/dmj2x) on this issue. Thanks for your feedback we will definitely use this information. To clarify, if an upload is attempted and there is no space, you want the upload button disabled? Or should they be able to try to upload and get an error message?

**Comment by foosel** [comment]
Disabling the upload button altogether is tricky, since you don't know if a file will still "fit" if you haven't selected it yet (which you'll do through the upload button). So I'd say, just make sure the error message is a meaningful one, that should already give enough information to the user to take meaningful measures in order to rectify the situation.

**Comment by theosmacmillan** [comment]
That makes sense. Another question: we were able to show zero space by replacing 
`return`jsonify(files=files, free=usage.free, total=usage.total)` 
with 
`return jsonify(files=files, free=0, total=usage.total)`
 into the "readGcodeFiles()" function in the file: 
'OctoPrint/src/octoprint/server/api/files.py'

However this doesn't change the actual available disk space in OctoPrint. Is there a way to actually point the disk space to zero space? Just so we can recreate that scenario?

**Comment by markwal** [comment]
My guess is find the place where usage.free is set and set it to 0 instead.

**Comment by theosmacmillan** [comment]
We have done this, and it doesn't actually change the available space. But we are working on finding the uploads folder which should have a size that we can set to 0.

**Comment by theosmacmillan** [comment]
Is there any way we could get a simple lifecycle of an uploaded file? Where it is stored temporarily, etc?

**Comment by foosel** [comment]
It's stored temporarily in the system's temporary folder (Usually `/tmp` on Linux, `%TEMP%` on Windows) during handling of the request by the upload handler of Tornado (`octoprint.server.util.tornado.UploadStorageFallbackHandler`). Then that temporary path gets passed on to the `/files` API handler (`octoprint.server.api.files#uploadGcodeFile`), which in turn hands over the path to the FileManager's `add_file` method. That's when the file is moved over to the final storage location, which is usually `~/.octoprint/uploads` (Linux) or `%APPDATA%/OctoPrint/uploads` (Windows), unless you configured a different path under Settings > Folders.

So, very basic ASCII diagram:

```
Browser --request--> Tornado --temporary path & size--> flask / files api handler --path--> file manager

                       vvv                                                                      vvv

                    temporary                                                             move to upload
                      file                                                                    folder
```

**Comment by foosel** [comment]
@theosmacmillan @dmj2x what's the status of this?

**Comment by theosmacmillan** [comment]
I have our solution (outlined here: https://www.youtube.com/watch?v=3phU1yKKQcw) in my local files. When we had finished the code for it, the semester was just ending and we did not make a pull request because we heard nothing would get approved. I'll try to get it together and merge it.

**Comment by foosel** [comment]
Assuming this is up for grabs again :/

**Comment by jromero** [comment]
@foosel looks like this issue should be closed per https://github.com/OctoPrint/OctoPrint/pull/4588
